driver-core/src/test/unit/com/mongodb/internal/connection/AbstractServerDiscoveryAndMonitoringTest.java

```diff
@@ -73,6 +78,43 @@ protected void applyResponse(final BsonArray response) {
         factory.sendNotification(serverAddress, serverDescription);
     }
 
+    protected void applyApplicationError(final BsonDocument applicationError) {
+        ServerAddress serverAddress = new ServerAddress(applicationError.getString("address").getValue());
+        int errorGeneration = applicationError.getNumber("generation",
+                new BsonInt32(((DefaultServer) getCluster().getServer(serverAddress)).getConnectionPool().getGeneration())).intValue();
+        int maxWireVersion = applicationError.getNumber("maxWireVersion").intValue();
+        String when = applicationError.getString("when").getValue();
+        String type = applicationError.getString("type").getValue();
+
+        ClusterableServer server = cluster.getServer(serverAddress);
+        RuntimeException exception;
+
+        switch (type) {
+            case "command":
+                exception = getCommandFailureException(applicationError.getDocument("response"), serverAddress);
+                break;
+            case "network":
+                exception = new MongoSocketReadException("Read error", serverAddress, new IOException());
+                break;
+            case "timeout":
+                exception = new MongoSocketReadTimeoutException("Read timeout error", serverAddress, new IOException());
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported application error type: " + type);
+        }
+
+        switch (when) {
+            case "beforeHandshakeCompletes":
+                server.invalidate(errorGeneration);
```

Comment:
Note that this is testing the logic at the level of the invalidate methods.  It's assuming that `com.mongodb.internal.connection.DefaultServer.DefaultServerProtocolExecutor` is calling invalidate properly.

It turned out not to be feasible to enter into the logic in DefaultServerProtocolExecutor itself.